### PR TITLE
Fsrepo cleanup

### PR DESCRIPTION
**Pull request by tv42**

### CODE REVIEW

**Comment by jbenet** [comment]
@tv42 yay less code! :D 

sharness tests fail. try:

```sh
make test

# ok

cd test/sharness
make
```

File: repo/fsrepo/fsrepo.go [hunk 1]
```diff
@@ -164,30 +180,17 @@ func Init(repoPath string, conf *config.Config) error {
 
 // Remove recursively removes the FSRepo at |path|.
 func Remove(repoPath string) error {
-	repoPath = path.Clean(repoPath)
```
Comment (jbenet) [inline repo/fsrepo/fsrepo.go]: why aren't we cleaning the path anymore?

File: repo/fsrepo/fsrepo.go [hunk 2]
```diff
@@ -34,55 +31,74 @@ var (
...
-	//
-	// If an operation is used when repo is Open and the operation does not
-	// change the repo's state, the package lock does not need to be acquired.
-	openersCounter *counter.Openers
-
-	// protects dsOpenersCounter and datastores
-	dsLock           sync.Mutex
-	dsOpenersCounter *counter.Openers
-	datastores       map[string]ds2.ThreadSafeDatastoreCloser
+	packageLock sync.Mutex
```
Comment (jbenet) [inline repo/fsrepo/fsrepo.go]: why is the `packageLock` needed? i remember knowing once.

File: repo/fsrepo/fsrepo.go [hunk 3]
```diff
@@ -34,55 +31,74 @@ var (
...
+	r := &FSRepo{
+		path: expPath,
+	}
+
+	lockfile, err := lockfile.Lock(r.path)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if lockfile != nil {
```
Comment (jbenet) [inline repo/fsrepo/fsrepo.go]: when can `lockfile` be nil?

File: repo/fsrepo/fsrepo.go (hunk 2 above)
Comment (tv42) [inline repo/fsrepo/fsrepo.go]: I'm not quite sure, and it's no longer not in my way, so I left it alone for now. I feel like it should go as a struct field if anything, but then there are a few non-method functions that use it.

File: repo/fsrepo/fsrepo.go (hunk 3 above)
Comment (tv42) [inline repo/fsrepo/fsrepo.go]: I set it to nil when we reach the point of success, so the defer unlocks on errors, but leaves lock alive on success. It's a bit too convoluted, but I couldn't come up with a nicer way off the top of my head. At least it holds the lock at the right time now.

File: repo/fsrepo/fsrepo.go (hunk 1 above)
Comment (tv42) [inline repo/fsrepo/fsrepo.go]: It went inline in the TildeExpansion call. I can yank it out to a separate line if you prefer that, but now all the path manipulation is very tightly grouped and I felt it reads well enough. Just a misleading diff.

```go
	expPath, err := u.TildeExpansion(path.Clean(repoPath))
	if err != nil {
		return nil, err
	}

	r := &FSRepo{
		path: expPath,
	}
```

File: repo/fsrepo/fsrepo.go (hunk 3 above)
Comment (jbenet) [inline repo/fsrepo/fsrepo.go]: @tv42 maybe using another variable will make it more readily understandable?

```go
succeeded = false
defer func() {
	if !succeeded {
		lockfile.Close()
	}
}
....
succeeded = true
return r, nil
```

File: repo/fsrepo/fsrepo.go (hunk 1 above)
Comment (jbenet) [inline repo/fsrepo/fsrepo.go]: @tv42 ahh didnt see that there. But, wait, that means that:

`fsrepo.Open(p)` and `fsrepo.Remove(p)` may do different things depending on how the user's call them or `p`'s value. i think they should probably be symmetric, both sanitize input.

File: repo/fsrepo/fsrepo.go [hunk 4]
```diff
@@ -164,30 +180,17 @@ func Init(repoPath string, conf *config.Config) error {
...
-
-	// packageLock must be held to check the number of openers.
-	packageLock.Lock()
-	defer packageLock.Unlock()
+	// TODO replace this with the "api" file
+	// https://github.com/ipfs/specs/tree/master/repo/fs-repo
 
 	// NB: the lock is only held when repos are Open
-	return lockfile.Locked(repoPath) && openersCounter.NumOpeners(repoPath) == 0
+	return lockfile.Locked(repoPath)
```
Comment (jbenet) [inline repo/fsrepo/fsrepo.go]: input should probably be cleaned here too, since the func is exported

File: repo/fsrepo/fsrepo.go (hunk 1 above)
Comment (jbenet) [inline repo/fsrepo/fsrepo.go]: (i.e. since func is exported, they should all behave the same way about paths)

File: repo/fsrepo/fsrepo_test.go [hunk 5]
```diff
@@ -30,31 +30,7 @@ func TestInitIdempotence(t *testing.T) {
 func TestRemove(t *testing.T) {
 	t.Parallel()
 	path := testRepoPath("foo", t)
-	assert.Nil(Remove(path), t, "should be able to remove after closed")
-}
-
-func TestCannotRemoveIfOpen(t *testing.T) {
```
Comment (jbenet) [inline repo/fsrepo/fsrepo_test.go]: wait, why did this test go away? one proc may have the repo open-- another proc must not be able to remove it.

File: repo/fsrepo/fsrepo_test.go [hunk 6]
```diff
@@ -125,18 +101,3 @@ func TestDatastorePersistsFromRepoToRepo(t *testing.T) {
 	assert.Nil(r2.Close(), t)
 	assert.True(bytes.Compare(expected, actual) == 0, t, "data should match")
 }
-
-func TestOpenMoreThanOnceInSameProcess(t *testing.T) {
```
Comment (jbenet) [inline repo/fsrepo/fsrepo_test.go]: prob should bring back this test after the memoizing thing

File: repo/fsrepo/fsrepo.go (hunk 1 above)
Comment (tv42) [inline repo/fsrepo/fsrepo.go]: `path.Clean` should never change the filesystem semantics of a path, but I'll put the cleaning back in just to make it read the same way.

File: repo/fsrepo/fsrepo.go (hunk 4 above)
Comment (tv42) [inline repo/fsrepo/fsrepo.go]: Restored that line.

File: repo/fsrepo/fsrepo_test.go (hunk 5 above)
Comment (tv42) [inline repo/fsrepo/fsrepo_test.go]: The test is about the process *itself* holding it open. Nobody does that, and it just sounds like a very bad bit of programming. `Remove` did not check for other processes before this commit, and isn't checking now -- the calling `ipfs init` does that.

Comment (tv42) [inline repo/fsrepo/fsrepo_test.go]: Well, actually, `ipfs init` doesn't check for other running processes, I remembered that wrong. What it does is test for existence, and normally bail out with an error. With `--force`, it instead removes the repo, and initializes it from scratch. 

That doesn't change my understanding though, and this commit did not change real behavior. The `ipfs init` itself didn't hold it open, so the old refcounting saw nothing special there.

File: repo/fsrepo/fsrepo.go (hunk 3 above)
Comment (tv42) [inline repo/fsrepo/fsrepo.go]: New version of this piggybacks on `err`, and avoids the `lockfile` variable completely, how's that:

```go
       defer func() {
               // unlock on error, leave it locked on success
               if err != nil {
                       r.lockfile.Close()
               }
       }()
```

Comment (tv42) [inline repo/fsrepo/fsrepo.go]: Okay I already hurt myself with needing a named return, doing it the very explicit way ;)

File: repo/fsrepo/fsrepo_test.go (hunk 6 above)
Comment (tv42) [inline repo/fsrepo/fsrepo_test.go]: Good catch, reordering commits so it never goes away.

**Comment by tv42** [comment]
@jbenet All nits responded to and passes sharness.

File: repo/fsrepo/fsrepo_test.go (hunk 5 above)
Comment (jbenet) [inline repo/fsrepo/fsrepo_test.go]: Yeah, you're right. the test's name was misleading. It didnt work across procs.

I think it's probably still a good idea to disable remove if another proc is holding the lock-- but that's not what this test did. We can do that in future PR.

**Comment by jbenet** [comment]
@tv42 awesome, everything LGTM.

One thing: we make sure that all our commits are passing, so that we can `git bisect` problems easily. (i.e. rebase bad things out).

We have a test that is supposed to check this: 
- https://build.protocol-dev.com/job/ipfs_test_all_commits_on_branch/1126/console
which passed, but looking at the console output, it didnt seem to run everything.

Running `make test_all_commits` in a linux machine manually just to make sure. will post outcome here shortly. if all good will merge. (found a bug in `make_test_all_commits` on linux too #429)

**Comment by jbenet** [comment]
checked on the results-- all passed fine :+1:

**Pull request merged**